src/sage/combinat/k_regular_sequence.py

```diff
@@ -918,6 +1026,265 @@ def _n_to_index_(self, n):
         except OverflowError:
             raise ValueError('value {} of index is negative'.format(n)) from None
 
+    def guess(self, f, n_max=100, max_dimension=10, sequence=None):
+        r"""
+        Guess a `k`-regular sequence of `(f(n))_{n\geq0}`.
+
+        INPUT:
+
+        - ``f`` -- a function (callable) which determines the sequence.
+          It takes nonnegative integers as an input
+
+        - ``n_max`` -- (default: ``100``) a positive integer. The resulting
+          `k`-regular sequence coincides with `f` on the first ``n_max``
+          terms
+
+        - ``max_dimension`` -- (default: ``10``) a positive integer specifying
+          the maxium dimension which is tried when guessing the sequence
+
+        - ``sequence`` -- (default: ``None``) a `k`-regular sequence used
+          for bootstrapping this guessing
+
+        OUTPUT:
+
+        A :class:`kRegularSequence`
+
+        EXAMPLES:
+
+        Binary sum of digits::
+
+            sage: @cached_function
+            ....: def s(n):
+            ....:     if n == 0:
+            ....:         return 0
+            ....:     return s(n//2) + ZZ(is_odd(n))
+            sage: all(s(n) == sum(n.digits(2)) for n in srange(10))
+            True
+            sage: [s(n) for n in srange(10)]
+            [0, 1, 1, 2, 1, 2, 2, 3, 1, 2]
+
+        Variant 1::
+
+            sage: Seq2 = kRegularSequenceSpace(2, ZZ)
+            sage: import logging
+            sage: logging.basicConfig(level=logging.INFO)
+            sage: S1 = Seq2.guess(s)
+            INFO:...:including f_{1*m+0}
+            INFO:...:M_0: f_{2*m+0} = (1) * X_m
+            INFO:...:including f_{2*m+1}
+            INFO:...:M_1: f_{2*m+1} = (0, 1) * X_m
+            INFO:...:M_0: f_{4*m+1} = (0, 1) * X_m
+            INFO:...:M_1: f_{4*m+3} = (-1, 2) * X_m
+            sage: S1
+            2-regular sequence 0, 1, 1, 2, 1, 2, 2, 3, 1, 2, ...
+            sage: S1.mu[0], S1.mu[1], S1.left, S1.right
+            (
+            [1 0]  [ 0  1]
+            [0 1], [-1  2], (1, 0), (0, 1)
+            )
+
+            sage: from importlib import reload
+            sage: logging.shutdown(); _ = reload(logging)
+
+        Variant 2::
+
+            sage: C = Seq2((Matrix([[1]]), Matrix([[1]])), vector([1]), vector([1])); C
+            2-regular sequence 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
+            sage: S2 = Seq2.guess(s, sequence=C)
+            sage: S2
+            2-regular sequence 0, 1, 1, 2, 1, 2, 2, 3, 1, 2, ...
+            sage: S2.mu[0], S2.mu[1], S2.left, S2.right
+            (
+            [1 0]  [1 0]
+            [0 1], [1 1], (0, 1), (1, 0)
+            )
```

Comment:
I suggest to compare the outputs of Variant 1 and Variant 2 (that they yield the same sequence).

https://github.com/sagemath/sage/issues/21204#issuecomment-1417820841 Item 7